### PR TITLE
bpo-43950: Initial base implementation for PEP 657

### PR DESCRIPTION
**Pull request by pablogsal**
⚠️  This is not the full feature and contains no optimizations for the time being, just the base implementation to avoid merge conflicts in the future ⚠️ 

**All the internal structures, and data representation are subjected to change in future PRs as we add optimizations**

This PR is part of PEP 657 and augments the compiler to emit ending
line numbers as well as starting and ending columns from the AST
into compiled code objects. This allows bytecodes to be correlated
to the exact source code ranges that generated them.

This information is made available through the following public APIs:

* The `co_positions` method on code objects.
* The C API function `PyCode_Addr2Location`.

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>
Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>
Co-authored-by: Ammar Asker <ammar@ammaraskar.com>

https://bugs.python.org/issue43950

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -1065,12 +1065,6 @@ static PyGetSetDef type_getsets[] = {
 static PyObject *
 type_repr(PyTypeObject *type)
 {
-    if (type->tp_name == NULL) {
```
Comment (ammaraskar) [inline Objects/typeobject.c]: I think this accidentally got caught up in the merge

File: Python/traceback.c [hunk 2]
```diff
@@ -926,4 +1008,3 @@ _Py_DumpTracebackThreads(int fd, PyInterpreterState *interp,
 
     return NULL;
 }
-
```
Comment (ammaraskar) [inline Python/traceback.c]: Aah applying the github suggestion wiped this blank line.

Comment (isidentical) [reply Python/traceback.c]: Probably my bad, it was a rough merge with all the stuff changed. I think you should have commit privs on colnotab, so feel free to fix these irrelevant stuff in commit

**Comment by isidentical** [comment]
A couple doctest failures FYI: https://travis-ci.com/github/python/cpython/jobs/520228793

**Review by isidentical — APPROVED** [review]

File: Include/cpython/code.h [hunk 3]
```diff
@@ -165,6 +172,12 @@ PyCode_NewEmpty(const char *filename, const char *funcname, int firstlineno);
    in this code object.  If you just need the line number of a frame,
    use PyFrame_GetLineNumber() instead. */
 PyAPI_FUNC(int) PyCode_Addr2Line(PyCodeObject *, int);
+/* Return the ending source code line number from a bytecode index. */
+PyAPI_FUNC(int) PyCode_Addr2EndLine(PyCodeObject*, int);
+/* Return the starting source code column offset from a bytecode index. */
+PyAPI_FUNC(int) PyCode_Addr2Offset(PyCodeObject*, int);
+/* Return the ending source code column offset from a bytecode index. */
+PyAPI_FUNC(int) PyCode_Addr2EndOffset(PyCodeObject*, int);
```
Comment (Fidget-Spinner) [inline Include/cpython/code.h]: One question, doesn't this expose these functions as part of the public C API? Is that intentional?

**Comment by markshannon** [comment]
Instead of adding three additional location fields to each instruction, and continually update the equivalent fields in the scope struct, why not change `i_lineno` to `i_ast_node` to hold the AST node with the matching location?

Doesn't PEP 657 require that the locations match those of the AST exactly?

File: Include/cpython/code.h (hunk 3 above)
Comment (encukou) [reply Include/cpython/code.h]: It doesn't: `Include/cpython/*` is ~~internal~~ *not limited* API.

Comment (Fidget-Spinner) [reply Include/cpython/code.h]: Ohh right, I forgot that for ``code.h`` in ``Python.h``, it's the one in ``Include/*``. Sorry and thanks for clarification Petr.

**Comment by pablogsal** [comment]
> Instead of adding three additional location fields to each instruction, and continually update the equivalent fields in the scope struct, why not change `i_lineno` to `i_ast_node` to hold the AST node with the matching location?

That would be basically the same because you are modifiying these fields in the compilation and assembler steps. For example:

```
...
b->b_instr[i].i_lineno = prev_lineno;
...
```

that means that all instructions will share the same ast node, so changes to the lineno of one instruction will affect others, which is a big problem. So either we start adding a bunch of macros to get and set to handle the default case a NULL pointer for `i_ast_node` or we would end copiying the ast node location structure, which is almost exactly what we are doing currently.

**Comment by pablogsal** [comment]
@isidentical By the way, it seems that there are more places where we are setting `lineno = -1` that we didn't update to also account for the new fields

File: Include/cpython/code.h (hunk 3 above)
Comment (pablogsal) [reply Include/cpython/code.h]: @encukou  Is internal API but unless I am mistaken these are public functions. For instance, this file contains `PyCode_New` which is definitely public

Comment (pablogsal) [reply Include/cpython/code.h]: But in any case, yes, there is a new C-API function here (see the pep)

**Comment by ammaraskar** [comment]
@terryjreedy Adding you as a reviewer as requested. You might not be too interested in this part, it's just exposing the lower level offset data but the tests might give you sort of an idea on how they can be used from Python. The follow up PRs will add the `traceback.py` changes that you might be more interested in.

File: Lib/test/test_compile.py [hunk 4]
```diff
@@ -985,6 +986,118 @@ def if_else_break():
...
+            if end_line is not None:
+                self.assertIn(end_line, end_lines)
+            if col is not None:
+                self.assertIn(col, columns)
+            if end_col is not None:
+                self.assertIn(end_col, end_columns)
+
+        return code, ast_tree
+
+    def assertOpCodeSourcePositionIs(self, code, opcode,
```
Comment (terryjreedy) [inline Lib/test/test_compile.py]: Here and below...
```suggestion
    def assertOpcodeSourcePositionIs(self, code, opcode,
```
is more readable to me.

File: Lib/test/test_compile.py [hunk 5]
```diff
@@ -985,6 +986,118 @@ def if_else_break():
...
+
+    def test_compiles_to_extended_op_arg(self):
+        # Make sure we still have valid positions when the code compiles to an
+        # EXTENDED_ARG by performing a loop which needs a JUMP_ABSOLUTE after
+        # a bunch of opcodes.
+        snippet = "x = x\n" * 10_000
+        snippet += "while x != 0:\n"
+        snippet += "  x -= 1\n"
+        snippet += "while x != 0:\n"
+        snippet += "  x +=  1\n"
```
Comment (terryjreedy) [inline Lib/test/test_compile.py]: Copy 10000 lines just once.  To me, easier to read also.
```suggestion
        snippet += ("while x != 0:\n"
                    "  x -= 1\n"
                    "while x != 0:\n"
                    "  x +=  1\n"
                   )
```

**Review by terryjreedy — COMMENTED** [review]
If I understand, the only officially visible change is the addition of code.co_positions, which returns an iterable of (start_line, end_line, start_column, end_column) tuples, where column numbers are 1-based.  Suggested title after bpo tag: "Add code.co_positions (PEP 657)" or "Add PEP-657 code.co_positions".

I presume a doc change will come.  I will review it for readability if pinged.

I read the tests and have two suggestions.

**Comment by ammaraskar** [comment]
Amended the commit message as per Terry's recommendation and rebased https://github.com/python/cpython/pull/26958 on top of the latest changes, tests are still passing.

**Pull request merged**